module/documents/active-effect.mjs

```diff
@@ -333,6 +347,49 @@ export default class ActiveEffect5e extends ActiveEffect {
 
   /* -------------------------------------------- */
 
+  /**
+   * Prepare effect favorite data.
+   * @returns {Promise<FavoriteData5e>}
+   */
+  async getFavoriteData() {
+    return {
+      img: this.img,
+      title: this.name,
+      subtitle: this.duration.remaining ? this.duration.label : "",
+      toggle: !this.disabled,
+      suppressed: this.isSuppressed
+    };
+  }
+
+  /* -------------------------------------------- */
+  /*  Socket Event Handlers                       */
+  /* -------------------------------------------- */
+
+  /** @inheritDoc */
+  async _preCreate(data, options, user) {
+    if ( await super._preCreate(data, options, user) === false ) return false;
+    if ( options.keepOrigin === false ) this.updateSource({ origin: this.parent.uuid });
+
+    // Enchantments cannot be added directly to actors
+    if ( (this.getFlag("dnd5e", "type") === "enchantment") && (this.parent instanceof Actor) ) {
+      ui.notifications.error("DND5E.Enchantment.Warning.NotOnActor", { localize: true });
+      return false;
+    }
+
+    if ( !this.isAppliedEnchantment ) return;
+
+    // Otherwise validate against the enchantment's restraints on the origin item
+    const origin = await fromUuid(this.origin);
+    const errors = origin?.system.enchantment?.canEnchant(this.parent);
+    this.updateSource({ disabled: false });
+    if ( !errors?.length ) return;
+
+    errors.forEach(err => ui.notifications.error(err.message));
```

Comment:
Don't feel great about putting UI notifications inside API code. These should probably be logged to the console, and we should spawn UI notifications at the place the user interaction happens. For enchantments I guess that would be in the Item sheet drop handler, and potentially additional places in the future when we add more ways to enchant Items.